src/Cassandra/Compression/LZ4Compressor.cs

```diff
@@ -25,13 +25,39 @@ internal class LZ4Compressor : IFrameCompressor
         public Stream Decompress(Stream stream)
         {
             var buffer = Utils.ReadAllBytes(stream, 0);
+            if (buffer.Length < 4)
+            {
+                throw new CorruptionException("Corrupt literal length");
+            }
+
             var outputLengthBytes = new byte[4];
             Buffer.BlockCopy(buffer, 0, outputLengthBytes, 0, 4);
             Array.Reverse(outputLengthBytes);
             var outputLength = BitConverter.ToInt32(outputLengthBytes, 0);
-            var decompressStream = new MemoryStream(LZ4.LZ4Codec.Decode(buffer, 4, buffer.Length - 4, outputLength), 0, outputLength, false, true);
+            var outputBuffer = new byte[outputLength];
+            var uncompressedSize = LZ4Codec.Decode(buffer, 4, buffer.Length - 4, outputBuffer, 0, outputLength);
+            if (uncompressedSize < 0)
+            {
+                throw new CorruptionException("Can't decode LZ4 bytes");
+            }
+
+            if (outputLength != uncompressedSize)
+            {
+                throw new CorruptionException(string.Format("Recorded length is {0} bytes but actual length after decompression is {1} bytes ",
+                    outputLength,
+                    uncompressedSize));
+            }
+
+            var decompressStream = new MemoryStream(outputBuffer, 0, outputLength, false, true);
             return decompressStream;
         }
+
+        private class CorruptionException : Exception
```

Comment:
I would honestly just use `DriverInternalError` here instead of creating a new exception

Reply:
Agreed, changed.